openscm/timeseries_converter.py

```diff
@@ -0,0 +1,364 @@
+this module. A thorough explaination of the procedure used is given in a dedicated
+`Jupyter Notebook
+<https://github.com/openclimatedata/openscm/blob/master/notebooks/timeseriess.ipynb>`_.
+"""
+
+from enum import Enum
+from typing import Tuple, Callable
+
+
+import numpy as np
+import scipy.interpolate as interpolate
+import scipy.integrate as integrate
+
+
+from .errors import InsufficientDataError
+from .parameters import ParameterType
+
+# pylint: disable=too-many-arguments
+
+
+class ExtrapolationType(Enum):
+    """
+    Extrapolation type.
+    """
+
+    NONE = 0
+    LINEAR = 1
+    # TODO support CUBIC = 3
+
+
+class InterpolationType(Enum):
+    """
+    Interpolation type.
+    """
+
+    LINEAR = 1
+    # TODO support CUBIC = 3
+
+
+def create_time_points(
+    start_time: int, period_length: int, points_num: int, timeseries_type: ParameterType
+) -> np.ndarray:
+    """
+    Create time points for an equi-distant time series.
+
+    Parameters
+    ----------
+    start_time
+        First time point of the timeseries (seconds since ``1970-01-01 00:00:00``)
+    period_length
+        Period length (in seconds)
+    points_num
+        Length of timeseries
+    timeseries_type
+        Timeseries type
+
+    Returns
+    -------
+    np.ndarray
+        Array of the timeseries time points
+    """
+    return np.linspace(
+        start_time,
+        start_time + points_num * period_length,
+        points_num
+        + (
+            1  # +1 because for averages we need to give the full last interval
+            if timeseries_type == ParameterType.AVERAGE_TIMESERIES
+            else 0
+        ),
+    )
+
+
+def _calc_interval_averages(
+    continuous_representation: Callable[[float], float], target_intervals: np.ndarray
+) -> np.ndarray:
+    """
+    Calculate the interval averages of a continuous function.
+
+    Here interval average is calculated as the integral over the period divided by
+    the period length.
+
+    Parameters
+    ----------
+    continuous_representation
+        Continuous function from which to calculate the interval averages. Should be
+        calculated using
+        :func:`openscm.timeseries_converter.TimeseriesConverter._calc_continuous_representation`.
+    target_intervals
+        Intervals to calculate the average of.
+
+    Returns
+    -------
+    np.ndarray
+        Array of the interval/period averages
+    """
+
+    # TODO: numerical integration here could be very expensive
+    # TODO: update to include caching and/or analytic solutions depending on
+    # TODO:     interpolation choice
+    int_averages = [np.nan] * len(target_intervals[:-1])
+    for i, l in enumerate(target_intervals[:-1]):
+        u = target_intervals[i + 1]
+        y, _ = integrate.quad(continuous_representation, l, u)
+        int_averages[i] = y / (u - l)
+
+    return np.array(int_averages)
+
+
+def _calc_linearization_values(values: np.ndarray) -> np.ndarray:
```

Comment:
I think this can be deleted no? It doesn't get used anywhere and is a duplicate of calculate continuous?

Reply:
I lie ignore that comment. Perhaps we can rename somehow? `_calc_integral_preserving_linear_interpolation`?